pkg/generic/thrift/http.go

```diff
@@ -77,12 +110,87 @@ func (r *ReadHTTPResponse) SetBase64Binary(enable bool) {
 	r.base64Binary = enable
 }
 
+// SetUseRawBodyForHTTPResp ...
+func (r *ReadHTTPResponse) SetUseRawBodyForHTTPResp(useRawBodyForHTTPResp bool) {
+	r.useRawBodyForHTTPResp = useRawBodyForHTTPResp
+}
+
+// SetDynamicGo ...
+func (r *ReadHTTPResponse) SetDynamicGo(t2jBinaryConv *t2j.BinaryConv, msg remote.Message) {
+	r.t2jBinaryConv = *t2jBinaryConv
+	r.msg = msg
+	r.dynamicgoEnabled = true
+}
+
 // Read ...
 func (r *ReadHTTPResponse) Read(ctx context.Context, method string, in thrift.TProtocol) (interface{}, error) {
+	// fallback logic
+	if !r.dynamicgoEnabled {
+		return r.originalRead(ctx, method, in)
+	}
+
+	// dynamicgo logic
+	if r.msg.PayloadLen() == 0 {
+		return nil, perrors.NewProtocolErrorWithMsg("msg.PayloadLen should always be greater than zero")
+	}
+
+	tProt, ok := in.(*cthrift.BinaryProtocol)
+	if !ok {
+		return nil, perrors.NewProtocolErrorWithMsg("TProtocol should be BinaryProtocol")
+	}
+	mBeginLen := bthrift.Binary.MessageBeginLength(method, thrift.TMessageType(r.msg.MessageType()), r.msg.RPCInfo().Invocation().SeqID())
+	sName, err := in.ReadStructBegin()
+	if err != nil {
+		return nil, err
+	}
+	sBeginLen := bthrift.Binary.StructBeginLength(sName)
+	fName, typeId, id, err := in.ReadFieldBegin()
+	if err != nil {
+		return nil, err
+	}
+	fBeginLen := bthrift.Binary.FieldBeginLength(fName, typeId, id)
+	transBuf, err := tProt.ByteBuffer().ReadBinary(r.msg.PayloadLen() - mBeginLen - sBeginLen - fBeginLen - bthrift.Binary.MessageEndLength())
+	if err != nil {
+		return nil, err
+	}
+	fid := dthrift.FieldID(id)
+
+	resp := descriptor.NewHTTPResponse()
+	ctx = context.WithValue(ctx, conv.CtxKeyHTTPResponse, resp)
+	tyDsc := r.svc.DynamicGoDsc.Functions()[method].Response()
+	// json size is usually 2 times larger than equivalent thrift data
+	buf := mcache.Malloc(len(transBuf) * 2)[0:0]
+	defer mcache.Free(buf)
```

Comment:
buf's ownership has been moved to resp...why can it be free?

Reply:
just use normal `make([]byte ...)` here

Reply:
or if you are not sure about the capcity of buf, you can just use conv.Do() which gives out the buf

Reply:
fixed